Adding refactored docker_image module.

chouseknecht opened the pull request:
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

docker_image.py 
##### ANSIBLE VERSION

``````
ansible 2.1.0 (docker_testing 9bb069f873) last updated 2016/04/18 12:36:03 (GMT -400)
  lib/ansible/modules/core: (detached HEAD cf01087a30) last updated 2016/04/05 16:31:53 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 204b4bab56) last updated 2016/04/05 16:31:56 (GMT -400)
  config file =
  configured module search path = Default w/o overrides```

##### SUMMARY
The new docker_image module, completely replacing the current docker_image.
``````

Commit:
Adding refactored docker_image module.

gregdek commented:
Thanks @chouseknecht for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

bcoca commented on cloud/docker/docker_image.py:
```diff
@@ -1,457 +1,502 @@
+
+short_description: Manage docker images.
+
+version_added: "2.1.0"
```
actually, correct version is still 1.5 ...

bcoca commented on cloud/docker/docker_image.py:
```diff
@@ -1,457 +1,502 @@
+  config_path:
+    description:
+      - Path to the Docker CLI config file.
+    default: '~/.docker/config.json'
```
this should alias to 'path' for backwards compat

bcoca replied:
missing `required`

chouseknecht replied:
Actually config_path should be removed as path is defined further down. The correct thing is path, and path is the same as it was in the original module - the path to the docker build directory containing the Dockerfile.

bcoca commented:
in general:
- keep version_added of the existing directives if they were post 1.5
- make note that new options were added in 2.1

bcoca commented on cloud/docker/docker_image.py:
```diff
@@ -1,457 +1,502 @@
-    choices: [ "no", "encrypt", "verify" ]
-    version_added: "2.0"
-  tls_client_cert:
+  push:
```
this could just be infered when 'present' AND name has repository or ~~repository_url could be present as it's own option~~ the existing `repository`.

bcoca commented on cloud/docker/docker_image.py:
```diff
@@ -1,457 +1,502 @@
+      - "Use with state present to always push an image to the registry. The image name must contain a repository
+        path and optionally a registry. For example: registry.ansible.com/user_a/repository"
+    default: false
+  pull:
```
how does this interact with cache?

chouseknecht replied:
I have no idea. docker-py does not have a 'cache' option for push.

bcoca commented on cloud/docker/docker_image.py:
```diff
@@ -1,457 +1,502 @@
-  tls_ca_cert:
+      - When building an image downloads any updates to the FROM image in Dockerfiles.
+    default: true
+  rm:
```
also merits interaction with cache explanation

chouseknecht replied:
Here is all that docker-py offers:  Downloads any updates to the FROM image in Dockerfiles.  I don't know how it interacts with cache. Seem like updating would invalidate the cache.

bcoca commented on cloud/docker/docker_image.py:
```diff
@@ -1,457 +1,502 @@
+      - Used to select an image when pulling. Will be added to the image when pushing, tagging or building. Defaults to
+       'latest' when pulling an image.
+    default: latest
+  use_tls:
```
I would separeate this use_tls= boolean and validate_certs=boolan (usless unless tls=true), to match what we have standardized towards in other modules

chouseknecht replied:
Not sure what you mean. The tls and tls_verify parameters are handled in the common module. use_tls is here just to be compatible with the old docker_image. 

It's a bit late to change tls and tls_verify now, if that's what you're suggesting. They match exactly what docker-py has, and they match the environment vars (DOCKER_TLS and DOCLER_TLS_VERIFY) used by Docker machine. They are confusing, yes. But, it's consistent with how docker does things.

bcoca replied:
There is consistency with docker and there is consistency with other Ansible modules, we prefer the latter. 

But given that this had already 'slipped in' we need to keep for backwards compatibility.

bcoca commented on cloud/docker/docker_image.py:
```diff
@@ -1,457 +1,502 @@
+                        image_tar = open(self.load_path, 'r')
+                        image_data = image_tar.read()
+                        image_tar.close()
+                    except Exception, exc:
```
since this is already not 2.4 compatible, use py3 'friendly' exceptions:

`Exception as exc:`

bcoca commented on cloud/docker/docker_image.py:
```diff
@@ -1,457 +1,502 @@
+                    name = "%s:%s" % (self.name, self.tag)
+
+                if not self.check_mode:
+                    self.results['actions'].append("Loaded image %s from %s" % (name, self.load_path))
```
would we not want to show which actions we 'would have' run if not in check mode?

Commit:
Updater per PR comments

Commit:
Fix unfriendly exeption handling.

chouseknecht commented:
- Removed the push parameter and inferred from the name passed to the push method. If the name is a path (i.e. it contains a '/'), then perform the push. I think that makes the module a little opinionated, since it determines when to push an image, but maybe that's OK.
- Fixed exceptions to be py3 friendly.
- Removed config_path parameter. It should not have been there and was not actually used. Correct parameter is path.
- Fixed placement of results['actions'].append() and results['changed']=True statements. Things are now consistent throughout, and actions is always defined and changed is always set regardless of check_mode.
- Not changing tls and tls_verify parameters at this juncture, if that's what the comments were suggesting.

chouseknecht commented:
The tests in ansible/docker repo were update to match removal of the 'push' option. The tests now assert for when a push is expected or not expected.

Commit:
Fix github credentials

Commit:
Clean up and fixes post API version 1.20 testing.

Commit:
Updated version_added on options to reflect options added for 2.1.

chouseknecht commented:
Updated version_added for options.

gregdek commented:
Thanks @chouseknecht for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]
